services/stats/collector/collector.go

```diff
@@ -0,0 +1,152 @@
+package collector
+
+import (
+	"runtime"
+	"time"
+)
+
+// GaugeFunc is an interface that implements the setting of a gauge value
+// in a stats system. It should be expected that key will contain multiple
+// parts separated by the '.' character in the form used by statsd (e.x.
+// "mem.heap.alloc")
+type GaugeFunc func(key string, val uint64)
+
+// Collector implements the periodic grabbing of informational data from the
+// runtime package and outputting the values to a GaugeFunc.
+type Collector struct {
+	// PauseDur represents the interval in between each set of stats output.
+	// Defaults to 10 seconds.
+	PauseDur time.Duration
+
+	// EnableCPU determines whether CPU statistics will be output. Defaults to true.
+	EnableCPU bool
+
+	// EnableMem determines whether memory statistics will be output. Defaults to true.
+	EnableMem bool
+
+	// EnableGC determines whether garbage collection statistics will be output. EnableMem
+	// must also be set to true for this to take affect. Defaults to true.
+	EnableGC bool
+
+	// Done, when closed, is used to signal Collector that is should stop collecting
+	// statistics and the Run function should return. If Done is set, upon shutdown
+	// all gauges will be sent a final zero value to reset their values to 0.
+	Done <-chan struct{}
+
+	gaugeFunc GaugeFunc
+}
+
+// New creates a new Collector that will periodically output statistics to gaugeFunc. It
+// will also set the values of the exported fields to the described defaults. The values
+// of the exported defaults can be changed at any point before Run is called.
+func New(gaugeFunc GaugeFunc) *Collector {
+	return &Collector{
+		PauseDur:  10 * time.Second,
+		EnableCPU: true,
+		EnableMem: true,
+		EnableGC:  true,
+		gaugeFunc: gaugeFunc,
```

Comment:
Add support for stopping collector. This can be done by closing **Done** channel in collector object. This function can be used by any module handling server shutdown to stop collector which will publish zero metrics.